File: ai_server/ai/vcf_search.py
"""VCF (Victory by Continuous Four) search for Gomoku.

The key technique that lifts pattern-based play from "Level 5" to actually
winning games. Pattern MCTS alone gets stuck in mutual-defense stalemates
because neither side ever creates an unstoppable attack. VCF finds forced
win sequences by exploring only moves that create a four-threat (which
opponent MUST block), so the search tree is narrow and deep.

Algorithm:
  1. Try every move that creates a four (half_four / open_four / five).
  2. If it's a five → won.
  3. If it's an open_four → won next move (opponent can only block 1 side).
  4. Otherwise (half_four), find the single cell opponent MUST play to
     block the five-threat, place opponent there, and recurse.
  5. If recursion finds a win, we have a VCF line.

Branching factor is typically 1-4 (few positions create fours), so depth-8
search is fast (~ms).
"""

from ai.game_logic import BOARD_SIZE, EMPTY, BLACK, WHITE, DIRECTIONS
from ai.pattern_eval import _count_consecutive, _scan_line


def _makes_five(board, r, c, player):
    """Does placing player at (r, c) make 5 in a row?"""
    for dr, dc in DIRECTIONS:
        pos_c, _ = _count_consecutive(board, r, c, dr, dc, player)
        neg_c, _ = _count_consecutive(board, r, c, -dr, -dc, player)
        if 1 + pos_c + neg_c >= 5:
            return True
    return False


def _four_info(board, r, c, player):
    """After placing player at (r,c), return info about four-threats created.

    Detects both consecutive AND gapped patterns:
      - Consecutive four: XXXX with open end(s)
      - Gapped five: XXX_X or XX_XX etc. (5 stones with 1 gap → must block gap)

    Returns (kind, block_cells) where:
      kind: 'five' | 'open_four' | 'half_four' | None
      block_cells: cells opponent must play to prevent five
    """
    best_kind = None
    block_cells = []
    for dr, dc in DIRECTIONS:
        pos_con, pos_gap, pos_open, pos_gap_cell = _scan_line(
            board, r, c, dr, dc, player)
        neg_con, neg_gap, neg_open, neg_gap_cell = _scan_line(
            board, r, c, -dr, -dc, player)

        # --- Consecutive patterns (original logic) ---
        total = 1 + pos_con + neg_con
        open_ends = pos_open + neg_open
        if total >= 5:
            return 'five', []
        if total == 4:
            ends = []
            if pos_open:
                ends.append((r + (pos_con + 1) * dr, c + (pos_con + 1) * dc))
            if neg_open:
                ends.append((r - (neg_con + 1) * dr, c - (neg_con + 1) * dc))
            if open_ends == 2:
                if best_kind != 'five':
                    best_kind = 'open_four'
                block_cells.extend(ends)
            elif open_ends == 1:
                if best_kind not in ('five', 'open_four'):
                    best_kind = 'half_four'
                block_cells.extend(ends)

        # --- Gapped patterns (split-four detection) ---
        # 4+ stones across one gap → filling gap makes five → must block.
        # This is equivalent to a half_four with the gap as the block cell.
        if pos_gap > 0 and pos_gap_cell is not None:
            gap_total = 1 + pos_con + neg_con + pos_gap
            if gap_total >= 4:
                if best_kind not in ('five', 'open_four'):
                    best_kind = 'half_four'
                block_cells.append(pos_gap_cell)
        if neg_gap > 0 and neg_gap_cell is not None:
            gap_total = 1 + pos_con + neg_con + neg_gap
            if gap_total >= 4:
                if best_kind not in ('five', 'open_four'):
                    best_kind = 'half_four'
                block_cells.append(neg_gap_cell)

    return best_kind, block_cells


def _candidate_moves(board, radius=1):
    """Nearby empty cells (within radius of any stone).

    Returns a sorted list (by row, col) so iteration order is deterministic
    — the Swift port depends on this for bit-exact diff testing, and
    reproducible self-play games benefit too.
    """
    cands = set()
    has_stone = False
    for r in range(BOARD_SIZE):
        for c in range(BOARD_SIZE):
            if board[r][c] != EMPTY:
                has_stone = True
                for dr in range(-radius, radius + 1):
                    for dc in range(-radius, radius + 1):
                        nr, nc = r + dr, c + dc
                        if 0 <= nr < BOARD_SIZE and 0 <= nc < BOARD_SIZE:
                            if board[nr][nc] == EMPTY:
                                cands.add((nr, nc))
    if not has_stone:
        return [(BOARD_SIZE // 2, BOARD_SIZE // 2)]
    return sorted(cands)


def find_vcf(board, attacker, max_depth=10, max_branch=8):
    """Search for a forced-win move sequence via continuous fours.

    Returns the first winning move (r, c) or None if no VCF found.
    board is mutated during search but restored before return.

    Defaults (depth=10, branch=8) are the "max effort" settings. For
    real-time play, callers can pass smaller values to limit time.
    """
    return _vcf_recurse(board, attacker, max_depth, max_branch)


def _vcf_recurse(board, attacker, depth, max_branch):
    if depth <= 0:
        return None
    defender = WHITE if attacker == BLACK else BLACK

    # Immediate win check
    for r, c in _candidate_moves(board, radius=1):
        if _makes_five(board, r, c, attacker):
            return (r, c)

    # Gather four-threat moves
    four_moves = []
    for r, c in _candidate_moves(board, radius=2):
        board[r][c] = attacker
        kind, blocks = _four_info(board, r, c, attacker)
        board[r][c] = EMPTY
        if kind in ('open_four', 'half_four'):
            # open_four is instantly winning (unless defender also wins)
            priority = 2 if kind == 'open_four' else 1
            four_moves.append((priority, r, c, kind, blocks))

    # Try most promising first
    four_moves.sort(key=lambda x: -x[0])
    four_moves = four_moves[:max_branch]

    for _, r, c, kind, blocks in four_moves:
        board[r][c] = attacker

        # Counter-threat check: can the defender win IMMEDIATELY (five in
        # one move)? If so, skip this move.
        # Note: we don't check for defender's open_four counter here because
        # doing so at every recursive node is O(candidates²) and makes VCF
        # unusably slow. The attacker's half_four is a force-win in 2, and
        # defender's open_four counter is also a force-win in 2 — the race
        # is symmetric, so we commit to our VCF and let the outer MCTS
        # tiebreak if needed.
        defender_wins = False
        for dr, dc in _candidate_moves(board, radius=1):
            if _makes_five(board, dr, dc, defender):
                defender_wins = True
                break

        if defender_wins:
            board[r][c] = EMPTY
            continue

        if kind == 'open_four':
            # Opponent can only block one end; we win next move
            board[r][c] = EMPTY
            return (r, c)

        # half_four: opponent must block the single open end
        # (there might be multiple equivalent blocks if 2+ directions made fours)
        if not blocks:
            board[r][c] = EMPTY
            continue

        # Try each forced block; if ALL lead to recursive VCF wins, we've won.
        # sorted() for deterministic iteration (matches Swift port).
        all_blocked = True
        for br, bc in sorted(set(blocks)):
            if board[br][bc] != EMPTY:
                continue
            board[br][bc] = defender

            # Does defender's block create their own 4-threat we must answer?
            # For simplicity, ignore counter-threats and just recurse.
            sub = _vcf_recurse(board, attacker, depth - 1, max_branch)

            board[br][bc] = EMPTY
            if sub is None:
                all_blocked = False
                break

        board[r][c] = EMPTY
        if all_blocked:
            return (r, c)

    return None

File: ios_plugin/Sources/VcfSearch.swift
import Foundation

// Port of ai_server/ai/vcf_search.py. Finds forced-win sequences made
// entirely of four-threats: the opponent is compelled to block (or lose),
// so branching stays ~1-4 and depth-10 search runs in milliseconds.
//
// Shape intentionally mirrors Python so the diff tester can compare
// output directly. Board is passed as `inout [Int8]` (flat row-major
// 15×15) and mutated during recursion, restored before each function
// returns — exactly what the Python code does with 2D lists.

public enum VcfSearch {

	public enum FourKind: Int {
		case five = 3, openFour = 2, halfFour = 1
	}

	public struct FourInfo {
		public let kind: FourKind?
		public let blockCells: [(row: Int, col: Int)]
	}

	/// Does placing `player` at (row, col) complete five in a row?
	public static func makesFive(
		board: [Int8], row: Int, col: Int, player: Int8
	) -> Bool {
		for (dr, dc) in GameLogic.directions {
			let pos = PatternEval.countConsecutive(
				board: board, row: row, col: col,
				dRow: dr, dCol: dc, player: player)
			let neg = PatternEval.countConsecutive(
				board: board, row: row, col: col,
				dRow: -dr, dCol: -dc, player: player)
			if 1 + pos.count + neg.count >= 5 {
				return true
			}
		}
		return false
	}

	/// After placing `player` at (row, col), describe any four-threats it
	/// creates (consecutive open/half fours AND gapped 4+ stones that a
	/// single opponent move must block to prevent a five). block_cells is
	/// the set of cells the defender has to play to stop the five.
	public static func fourInfo(
		board: [Int8], row: Int, col: Int, player: Int8
	) -> FourInfo {
		var bestKind: FourKind? = nil
		var blockCells: [(row: Int, col: Int)] = []

		for (dr, dc) in GameLogic.directions {
			let pos = PatternEval.scanLine(
				board: board, row: row, col: col,
				dRow: dr, dCol: dc, player: player)
			let neg = PatternEval.scanLine(
				board: board, row: row, col: col,
				dRow: -dr, dCol: -dc, player: player)

			// Consecutive patterns
			let total = 1 + pos.consecutive + neg.consecutive
			let openEnds = pos.endOpen + neg.endOpen
			if total >= 5 {
				return FourInfo(kind: .five, blockCells: [])
			}
			if total == 4 {
				var ends: [(row: Int, col: Int)] = []
				if pos.endOpen == 1 {
					ends.append((row + (pos.consecutive + 1) * dr,
					             col + (pos.consecutive + 1) * dc))
				}
				if neg.endOpen == 1 {
					ends.append((row - (neg.consecutive + 1) * dr,
					             col - (neg.consecutive + 1) * dc))
				}
				if openEnds == 2 {
					if bestKind != .five {
						bestKind = .openFour
					}
					blockCells.append(contentsOf: ends)
				} else if openEnds == 1 {
					if bestKind != .five && bestKind != .openFour {
						bestKind = .halfFour
					}
					blockCells.append(contentsOf: ends)
				}
			}

			// Gapped patterns — 4+ stones across one gap, filling the gap
			// makes five. Treated as half_four with the gap as block cell.
			if pos.gapStones > 0, let g = pos.gapCell {
				let gapTotal = 1 + pos.consecutive + neg.consecutive + pos.gapStones
				if gapTotal >= 4 {
					if bestKind != .five && bestKind != .openFour {
						bestKind = .halfFour
					}
					blockCells.append((g.row, g.col))
				}
			}
			if neg.gapStones > 0, let g = neg.gapCell {
				let gapTotal = 1 + pos.consecutive + neg.consecutive + neg.gapStones
				if gapTotal >= 4 {
					if bestKind != .five && bestKind != .openFour {
						bestKind = .halfFour
					}
					blockCells.append((g.row, g.col))
				}
			}
		}

		return FourInfo(kind: bestKind, blockCells: blockCells)
	}

	/// Empty cells within `radius` of any stone on the board. Returns the
	/// list sorted by (row, col) so iteration order matches Python's
	/// sorted(list(set(...))) — diff tests rely on this determinism.
	public static func candidateMoves(
		board: [Int8], radius: Int
	) -> [(row: Int, col: Int)] {
		let n = GameLogic.boardSize
		var packed = Set<Int>()
		var hasStone = false
		for r in 0..<n {
			for c in 0..<n {
				if board[GameLogic.idx(r, c)] != 0 {
					hasStone = true
					for dr in -radius...radius {
						for dc in -radius...radius {
							let nr = r + dr
							let nc = c + dc
							if nr >= 0 && nr < n && nc >= 0 && nc < n
									&& board[GameLogic.idx(nr, nc)] == 0 {
								packed.insert(nr * n + nc)
							}
						}
					}
				}
			}
		}
		if !hasStone {
			return [(n / 2, n / 2)]
		}
		return packed.sorted().map { ($0 / n, $0 % n) }
	}

	/// Search for a forced-win four sequence. Returns the first winning
	/// move (row, col) or nil. `board` is mutated during search but
	/// restored before return. Defaults mirror Python (depth 10, branch 8).
	public static func findVcf(
		board: inout [Int8], attacker: Int8,
		maxDepth: Int = 10, maxBranch: Int = 8,
		forbiddenEnabled: Bool = false
	) -> (row: Int, col: Int)? {
		return vcfRecurse(board: &board, attacker: attacker,
		                  depth: maxDepth, maxBranch: maxBranch,
		                  forbiddenEnabled: forbiddenEnabled)
	}

	static func isLegalMove(
		board: [Int8], row: Int, col: Int, player: Int8, forbiddenEnabled: Bool
	) -> Bool {
		if board[GameLogic.idx(row, col)] != 0 { return false }
		if !forbiddenEnabled || player != Stone.black.rawValue { return true }
		var tmpBoard = board
		return !RenjuForbidden.isForbiddenBlack(board: &tmpBoard, row: row, col: col)
	}

	static func legalFourThreat(
		board: [Int8], kind: FourKind?, blocks: [(row: Int, col: Int)],
		attacker: Int8, forbiddenEnabled: Bool
	) -> (kind: FourKind, blocks: [(row: Int, col: Int)])? {
		guard let threatKind = kind,
				threatKind == .openFour || threatKind == .halfFour else {
			return nil
		}
		if !forbiddenEnabled || attacker != Stone.black.rawValue {
			return (threatKind, blocks)
		}

		let legalBlocks = blocks.filter {
			isLegalMove(
				board: board, row: $0.row, col: $0.col, player: attacker,
				forbiddenEnabled: forbiddenEnabled)
		}
		let uniqueCount = Set(legalBlocks.map {
			$0.row * GameLogic.boardSize + $0.col
		}).count
		if uniqueCount == 0 { return nil }
		if threatKind == .openFour && uniqueCount < 2 {
			return (.halfFour, legalBlocks)
		}
		return (threatKind, legalBlocks)
	}

	private static func vcfRecurse(
		board: inout [Int8], attacker: Int8, depth: Int, maxBranch: Int,
		forbiddenEnabled: Bool
	) -> (row: Int, col: Int)? {
		if depth <= 0 { return nil }
		let defender: Int8 = attacker == 1 ? 2 : 1

		// Immediate five check.
		for (r, c) in candidateMoves(board: board, radius: 1) {
			if !isLegalMove(
					board: board, row: r, col: c, player: attacker,
					forbiddenEnabled: forbiddenEnabled) {
				continue
			}
			if makesFive(board: board, row: r, col: c, player: attacker) {
				return (r, c)
			}
		}

		// Collect four-threat moves, prioritised (open_four first).
		// priority 2 = open_four, 1 = half_four.
		var fourMoves: [(priority: Int, row: Int, col: Int,
		                 kind: FourKind, blocks: [(row: Int, col: Int)])] = []
		for (r, c) in candidateMoves(board: board, radius: 2) {
			if !isLegalMove(
					board: board, row: r, col: c, player: attacker,
					forbiddenEnabled: forbiddenEnabled) {
				continue
			}
			board[GameLogic.idx(r, c)] = attacker
			let info = fourInfo(board: board, row: r, col: c, player: attacker)
			let legalThreat = legalFourThreat(
				board: board, kind: info.kind, blocks: info.blockCells,
				attacker: attacker, forbiddenEnabled: forbiddenEnabled)
			board[GameLogic.idx(r, c)] = 0
			if let threat = legalThreat {
				let prio = threat.kind == .openFour ? 2 : 1
				fourMoves.append((prio, r, c, threat.kind, threat.blocks))
			}
		}

		// Sort desc by priority; stable tiebreak on (r, c) matches Python
		// since candidateMoves is already sorted.
		fourMoves.sort { $0.priority > $1.priority }
		if fourMoves.count > maxBranch {
			fourMoves = Array(fourMoves.prefix(maxBranch))
		}

		for (_, r, c, kind, blocks) in fourMoves {
			board[GameLogic.idx(r, c)] = attacker

			// Skip if defender wins immediately (five in one).
			var defenderWins = false
			for (dr, dc) in candidateMoves(board: board, radius: 1) {
				if !isLegalMove(
						board: board, row: dr, col: dc, player: defender,
						forbiddenEnabled: forbiddenEnabled) {
					continue
				}
				if makesFive(board: board, row: dr, col: dc, player: defender) {
					defenderWins = true
					break
				}
			}
			if defenderWins {
				board[GameLogic.idx(r, c)] = 0
				continue
			}

			if kind == .openFour {
				board[GameLogic.idx(r, c)] = 0
				return (r, c)
			}

			if blocks.isEmpty {
				board[GameLogic.idx(r, c)] = 0
				continue
			}

			// Try each forced block; if ALL recurse to a win, we've won.
			// Deduplicate blocks via packed int for Set semantics.
			var uniqBlocks = Set<Int>()
			for b in blocks { uniqBlocks.insert(b.row * GameLogic.boardSize + b.col) }

			var allBlocked = true
			for packed in uniqBlocks.sorted() {
				let br = packed / GameLogic.boardSize
				let bc = packed % GameLogic.boardSize
				if !isLegalMove(
						board: board, row: br, col: bc, player: defender,
						forbiddenEnabled: forbiddenEnabled) {
					continue
				}
				board[GameLogic.idx(br, bc)] = defender
				let sub = vcfRecurse(board: &board, attacker: attacker,
				                     depth: depth - 1, maxBranch: maxBranch,
				                     forbiddenEnabled: forbiddenEnabled)
				board[GameLogic.idx(br, bc)] = 0
				if sub == nil {
					allBlocked = false
					break
				}
			}

			board[GameLogic.idx(r, c)] = 0
			if allBlocked {
				return (r, c)
			}
		}

		return nil
	}
}
